File: src/main/java/com/nerdearla/workshop/service/SellerService.java
package com.nerdearla.workshop.service;

import com.nerdearla.workshop.client.SellerClient;
import com.nerdearla.workshop.model.Seller;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

@Service
public class SellerService {

    private final SellerClient sellerClient;

    public SellerService(@Autowired SellerClient sellerClient) {
        this.sellerClient = sellerClient;
    }

    // Llamada a service externo
    public Seller findSeller(String sellerId) {
        return sellerClient.getById(sellerId);
    }
}

File: src/main/java/com/nerdearla/workshop/service/SellerService.kt
package com.nerdearla.workshop.service

import com.nerdearla.workshop.dto.seller.Seller
import com.nerdearla.workshop.validator.SellerValidator
import org.springframework.stereotype.Service
import org.springframework.web.reactive.function.client.WebClient

@Service
class SellerService(
    private val validator: SellerValidator,
    private val webClient: WebClient
) {
    fun findSeller(sellerId: String) =
        webClient
            .getSellerById(sellerId)
        .also {
            validator.validate(it)
        }

    private fun WebClient.getSellerById(sellerId: String) =
        get()
            .uri { uriBuilder ->
                uriBuilder
                    .path("/sellers/{sellerId}")
                    .build(sellerId)
            }
            .retrieve().bodyToMono(Seller::class.java).block()!!
}


